types/underscore/underscore-tests.ts

```diff
@@ -726,8 +768,584 @@ interface ChainTypeExtractor {
+    // that stop occurring when the minimum version is moved to TS 3.1, so use assignability checks instead for the moment as a workaround
+    let result: (string | undefined)[] = _.map(nonIntersectingPropertiesArray, stringRecordProperty);
+    _(nonIntersectingPropertiesArray).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesArray).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    result = _.collect(nonIntersectingPropertiesArray, stringRecordProperty);
+    _(nonIntersectingPropertiesArray).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesArray).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    result = _.map(nonIntersectingPropertiesArray, stringRecordProperty);
+    _(nonIntersectingPropertiesArray).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesArray).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    result = _.collect(nonIntersectingPropertiesArray, stringRecordProperty);
+    _(nonIntersectingPropertiesArray).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesArray).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    result = _.map(nonIntersectingPropertiesDictionary, stringRecordProperty);
+    _(nonIntersectingPropertiesDictionary).map(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesDictionary).map(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+    result = _.collect(nonIntersectingPropertiesDictionary, stringRecordProperty);
+    _(nonIntersectingPropertiesDictionary).collect(stringRecordProperty); // $ExpectType (string | undefined)[]
+    extractChainTypes(_.chain(nonIntersectingPropertiesDictionary).collect(stringRecordProperty)); // $ExpectType ChainType<(string | undefined)[], string | undefined>
+
+    // property name iterator with type any (see #33479)
+    _.map(stringRecordArray as any, stringRecordProperty); // $ExpectType any[]
+    _(stringRecordArray as any).map(stringRecordProperty); // $ExpectType any[]
+    extractChainTypes(_.chain(stringRecordArray as any).map(stringRecordProperty)); // $ExpectType ChainType<any[], any>
+    _.collect(stringRecordArray as any, stringRecordProperty); // $ExpectType any[]
+    _(stringRecordArray as any).collect(stringRecordProperty); // $ExpectType any[]
+    extractChainTypes(_.chain(stringRecordArray as any).collect(stringRecordProperty)); // $ExpectType ChainType<any[], any>
```

Comment:
Each of these blocks by itself looks fairly clean, concise and understandable by itself, partly thanks to the discussion we had before. But it goes on for pages and I find myself zoning out. After one page I was already not trying anymore to see the semantic difference between subsequent blocks; just scrolling through and vaguely noticing that the blocks had slightly different size and shapes.

I'm afraid this is too much, purely due to the human factor. I think you have to do something about this. Is is it possible to have fewer tests, more documentation, more variation or a bit of all of the above?

Reply:
The behavior of partial object iterators didn't end up varying in interesting ways for more complex cases; I covered the same possibilities as other categories of tests just to be safe, but I could probably cut most of those tests out.

I could also probably cut out the variations of all tests that test the existence of the `context` parameter since it's not an interesting variation, that's probably not used a ton, future editors are probably not likely to accidentally remove it, and it'll be very easy to add back if anyone does accidentally remove it. That won't quite cut the whole spread of tests I wrote for all functions in half, but it'll probably reduce it by a quarter or so. I could instead cut out the tests that check for the optional-ness of that parameter, but I'd prefer to go the first route since as I mentioned I don't think context is used tons so I feel it's more important to test that not specifying it works correctly.

If I were to cut out [this tricky business](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r446717762), that would also reduce the set of tests for `map` and `pluck` fairly significantly since I wouldn't need to be testing a bunch of different type union cases. I could also leave the tricky stuff in place but test it much less thoroughly (more just testing that the concept works at all rather than checking to make sure the concept is properly implemented for all overloads of `map` and `pluck`) if you'd like since it really covers what I consider to be more of an edge case and probably doesn't need as thorough of testing as a result.

Reply:
I can probably also drop list vs. array tests and only test lists since arrays are just a specific type of list.

Reply:
> I could also probably cut out the variations of all tests that test the existence of the `context` parameter since it's not an interesting variation, it's probably not used a ton, future editors are probably not likely to accidentally remove it, and it'll be very easy to add back if anyone does accidentally remove it.

Maybe still leave a few in, just to have a sampling.

> If I were to cut out [this tricky business](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r446717762), that would also reduce the set of tests for `map` and `pluck` fairly significantly since I wouldn't need to be testing a bunch of different type union cases. I could also leave the tricky stuff in place but test it much less thoroughly (more just testing that the concept works at all rather than checking to make sure the concept is properly implemented for all ~overloads~ collection parameter variations of interest for `map` and `pluck`) if you'd like since it really covers what I consider to be more of an edge case and probably doesn't need as thorough of testing as a result.

Let's have a quick thought experiment. If you're doing test-driven development, you add a test (or a few) that demonstrates #33479 before actually changing the types to address the bug. Since you're fixing #33479, and this is a valuable thing to do (so I'm not in favor of cutting out the tricky business), I suggest leaving in at least that one test/those few tests, so that if the bug is reintroduced later, we will be able to detect it. I'm writing this in vague general terms because I didn't take the effort to look at the bug in depth.

> I can probably also drop list vs. array tests and only test lists since arrays are just a specific type of list.

👍 

On top of the above measures to slim down the number of tests, maybe it is also possible to start each block with a line comment to announce what distinguishes it from previous blocks.

Reply:
> Maybe still leave a few in, just to have a sampling.

👍 

> Let's have a quick thought experiment. If you're doing test-driven development, you add a test (or a few) that demonstrates #33479 before actually changing the types to address the bug. Since you're fixing #33479, and this is a valuable thing to do (so I'm not in favor of cutting out the tricky business), I suggest leaving in at least that one test/those few tests, so that if the bug is reintroduced later, we will be able to detect it. I'm writing this in vague general terms because I didn't take the effort to look at the bug in depth.

Sorry, I'm not fixing #33479 in this PR, just avoiding re-introducing it by keeping an overload that takes a `string` and results in `any[]`. The way this evolved is below.

1. I updated the overload of `map` that takes a `string` to take a `K extends keyof T` instead
2. I realized that that didn't work with type unions
3. I figured out the tricky stuff to make it work with type unions
4. I found #33479 and realized that there was no way to make the tricky stuff work with a `T` of `any` because as far as I know there's no constraint that can determine that one has an `any`, so I added a fallback overload that takes any `string` and yields `any[]`
5. I realized that because I had that fallback anyway, I could pull back on the tricky stuff if necessary but thought I'd include it just in case it was a better idea than I might have thought and discuss making it simpler in [this comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r446717762)

> On top of the above measures to slim down the number of tests, maybe it is also possible to start each block with a line comment to announce what distinguishes it from previous blocks.

I do have some comments above blocks that test different overloads of iteratee, but I don't have them above dictionary and list variations for overloads. I'll add comments on list vs. dictionary to be more consistent with block comments.

Reply:
So to be extra clear, here are the options as I see them for the tricky business + avoiding re-introducing #33479:

1. `(K extends UnionTypes<T>) => UnionPropertyTypes<T, K>[]` + `(string) => any[]` = Figures out the resulting type for both simple types and type unions and doesn't re-introduce #33479
2. `(K extends keyof T) => T[K][]` + `(string) => any[]` = Figures out the resulting type for simple types, falls back to `any[]` for union types, doesn't re-introduce #33479
3. `(string) => any[]` = `any[]` for all cases

I think option `1` is clever but probably unnecessarily complex since type unions are most likely an edge case and I think option `3` is not as helpful as I'd like to be, so I'd recommend option `2`. Admittedly what I probably should have done here given that statement is check in option `2` and then leave discussion around how we could do this extra-complex option `1` thing if folks thought it was useful rather than doing this the other way around, though I must admit that I was very curious about what typescript-bot would tell me about the performance of `1`.

Reply:
Thanks for clarifying, I finally oversee it!

I strongly agree with not going for option 3. I mildly agree with going for option 2, mostly because I find it hard to tell at this point whether option 1 is actually correct in all cases. Option 2 is oversimplifying union types, but at least not in a way that might be wrong.

Reply:
Yeah, I think I'll switch to option 2. Aside from `Type | undefined` I don't think type unions are likely to be super common, I agree that while I think I've covered all the cases I'm not 100% sure, not using the tricky business is probably more performant, and dropping it will help cut down on tests per your comment [here](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45763#discussion_r448595585). It can always be brought back later if folks really want it or if I or someone else finds a better way to deal with #33479.

Sorry for dragging you through all of the discussion on the more complicated approach and then dropping it, and thanks for taking a look at it!

Reply:
I've cut out most of the test variations that were discussed in this thread. With regards to sampling the `context` parameter, when I went to go do it I ended up not being super happy about it, so I left it in for the moment; let me know what you think about these tests in their current state and whether you still think more reduction is warranted.